modules/core/domain/src/main/java/org/rhq/core/domain/content/PackageVersion.java

```diff
@@ -139,7 +142,13 @@
         + "   AND pv.repoPackageVersions IS EMPTY " //
         + "   AND pv.installedPackages IS EMPTY " //
         + "   AND pv.installedPackageHistory IS EMPTY "),
-
+    @NamedQuery(name = PackageVersion.DELETE_BY_PKG_IF_NO_CONTENT_SOURCES_OR_REPOS, query = "DELETE PackageVersion pv "
+        + " WHERE pv.id NOT IN (SELECT pvcs.packageVersion.id " //
+        + "                       FROM PackageVersionContentSource pvcs) " //
+        + "   AND pv.repoPackageVersions IS EMPTY " //
+        + "   AND pv.installedPackages IS EMPTY " //
+        + "   AND pv.installedPackageHistory IS EMPTY " //
+        + "   AND pv.generalPackage.id IN ( :packageIds )"),
```

Comment:
My memory may be failing me but I'm not totally sure this is safe because BundleFile is, I think, also backed by PackageVersion.   Still, it may be OK because bundle packages and resource packages should likely be independent.